extensions/ql-vscode/src/data-extensions-editor/library.ts

```diff
@@ -0,0 +1,55 @@
+import { basename, extname } from "../common/path";
+
+// From the semver package using
+// const { re, t } = require("semver/internal/re");
+// console.log(re[t.LOOSE]);
+// Modifications:
+// - Added version named group which does not capture the v prefix
+// - Removed the ^ and $ anchors
+// - Made the minor and patch versions optional
+// This will match any semver string at the end of a larger string
+const semverRegex =
+  /[v=\s]*(?<version>([0-9]+)(\.([0-9]+)(?:\.([0-9]+)(?:-?((?:[0-9]+|\d*[a-zA-Z-][a-zA-Z0-9-]*)(?:\.(?:[0-9]+|\d*[a-zA-Z-][a-zA-Z0-9-]*))*))?(?:\+([0-9A-Za-z-]+(?:\.[0-9A-Za-z-]+)*))?)?)?)/g;
+
+export interface Library {
+  name: string;
+  version?: string;
+}
+
+export function parseLibraryFilename(filename: string): Library {
```

Comment:
The parsing often works, but it did get confused at least once on each of the repos I tried. It seems to mostly be going wrong when there's an extra suffix like a timestamp, `beta`, or `M2`. How hard would it be to cover these cases?

I'm not sure if we can make sure it's always perfect in every case because there will always be a library with a weird-enough version string, but we can certainly reduce the cases where a human looks at it and notices that it's off.

Here are some examples of the data extensions editor on a few different java projects:

- `jhy/jsoup`
    <img width="587" alt="Screenshot 2023-07-14 at 12 25 37" src="https://github.com/github/vscode-codeql/assets/3749000/e2add4ce-7ff4-4fd0-a599-a1137ec529f3">

- `google/gson`
    <img width="587" alt="Screenshot 2023-07-14 at 12 26 32" src="https://github.com/github/vscode-codeql/assets/3749000/15b0b2a1-414c-4d29-addd-a30bdae6bc4f">

- `apache/maven`
    <img width="635" alt="Screenshot 2023-07-14 at 12 28 08" src="https://github.com/github/vscode-codeql/assets/3749000/cde9bbdf-56f9-4b5c-97d5-ddf9b8426d99">

Reply:
I've made some changes to the regex that makes it more lenient in matching versions. It now works properly for all the examples in the screenshot and didn't regress in matching any of the other versions, so it should be better for matching versions overall.

Reply:
Thanks for making those changes. All the examples I just tried worked perfectly.